<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [31]:
import pandas as pd
import psycopg2

In [32]:
DBNAME = '....'
USER = '....'
PASSWORD = '....'
HOST = '....'
PORT = ....

In [33]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

**Задание 3.1 - Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies).** 

In [6]:
# текст запроса
query_3_1 = f'''
            select
                count(*) as total_vacancies
            from public.vacancies
            '''

In [7]:
total_vacancies = pd.read_sql_query(query_3_1, connection)
display(total_vacancies)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


total_vacancies
0            49197

**Задание 3.2 - Напишите запрос, который посчитает количество работодателей (таблица employers).** 

In [8]:
# текст запроса
query_3_2 = f'''
            select
                count(*) as total_employers
            from public.employers
            '''

In [9]:
total_employers = pd.read_sql_query(query_3_2, connection)
display(total_employers)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


total_employers
0            23501

**Задание 3.3 - Посчитате с помощью запроса количество регионов (таблица areas).**

In [10]:
# текст запроса
query_3_3 = f'''
            select
                count(*) as total_areas
            from public.areas
            '''

In [11]:
# результат запроса
total_areas = pd.read_sql_query(query_3_3, connection)
display(total_areas)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


total_areas
0         1362

**Задание 3.4 - Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).**

In [12]:
# текст запроса
query_3_4 = f'''
            select
                count(*) as total_industries
            from public.industries
            '''

In [13]:
# результат запроса
total_industries = pd.read_sql_query(query_3_4, connection)
display(total_industries)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


total_industries
0               294

***

### Выводы по результатам "Знакомства с данными" и предварительного анализа данных

База данных состоит из 5 таблиц связанных с собой посредством ключей:

1. База "Areas" - таблица справочник, которая хранит код города и и его название:
    * В таблице 1362 строки и 2 столбца, пустых строк нет;
    * Все значения уникальные, общее количество регионов / городов - 1362.
    
2. База "Employers" - таблица справочник со списком работодателей:
    * В таблице 23501 строка и 3 столбца, пустых строк нет;
    * Уникальных значений по ID - 23501, по названию - 23175. Есть одноименные компании, наибольшее повторение - 7, компания "Перспектива";
    * Общее количество работодателей - 23501.

3. База "Industries" - таблица-справочник вариантов сфер деятельности работодателей:
    * В таблице 294 строки и 2 столбца, пустых строк нет;
    * Все значения уникальны, общее количество сфер деятельности - 294.

4. База "Employers_industries" - дополнительная таблица показывающая связь между работодателями и сферами их деятельности:
    * В таблице 32333 строки и 2 столбца. пустых строк нет.

5. База "Vacancies" - таблица хранит в себе перечень вакансий:
    * В таблице 49197 строк и 10 столбцов. Пустые строки есть в 3 признаках: Ключевые навыки, ЗП минимум и ЗП максимум;
    * Уникальных значений по наименованию - 21223, самое распространенное наименование "Системный администратор" - 1761;
    * Общее количество вакансий в базе - 49197.

# Юнит 4. Детальный анализ вакансий

**Задание 4.1 - Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area). Отсортируйте по количеству вакансий в порядке убывания.**

In [26]:
# текст запроса
query_4_1 = f'''
            select
                a.name as area,
                count(*) as cnt
            from public.areas as a
            join public.vacancies as v on a.id = v.area_id
            group by area
            order by cnt desc
            limit 5
            '''

In [27]:
# результат запроса
total_vacancies_areas = pd.read_sql_query(query_4_1, connection)
display(total_vacancies_areas)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892

**Задание 4.2 - Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.**

In [17]:
# текст запроса
query_4_2 = f'''
            select
                count(*)
            from public.vacancies as v
            where salary_from is not null or salary_to is not null
            '''

In [18]:
# результат запроса
salary_from_to_is_not_null = pd.read_sql_query(query_4_2, connection)
display(salary_from_to_is_not_null)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


count
0  24073

**Задание 4.3 - Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.**

In [19]:
# текст запроса
query_4_3 = f'''
            select
                avg(salary_from) as avg_salary_from,
                avg(salary_to) as avg_salary_to
            from public.vacancies as v
            '''

In [20]:
# результат запроса
avarege_salary = pd.read_sql_query(query_4_3, connection)
display(round(avarege_salary))

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


avg_salary_from  avg_salary_to
0          71065.0       110537.0

### Дополнительный запрос 4.3.1 - средняя ЗП в части признака - опыт работы.

In [22]:
# текст запроса
query_4_3_1 = f'''
            select
                v.experience as experience,
                count(v.id) as cnt_vacancies,
                avg(salary_from) as avg_salary_from,
                avg(salary_to) as avg_salary_to
            from public.vacancies as v
            group by v.experience
            order by avg_salary_from desc
            '''

In [23]:
avarege_salary_experience = pd.read_sql_query(query_4_3_1, connection)
display(round(avarege_salary_experience))

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  cnt_vacancies  avg_salary_from  avg_salary_to
0         Более 6 лет           1337         150156.0       197819.0
1       От 3 до 6 лет          14511         112544.0       171974.0
2  От 1 года до 3 лет          26152          64456.0        97436.0
3           Нет опыта           7197          34699.0        56990.0

**Задание 4.4 - Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.**


In [21]:
# текст запроса
query_4_4 = f'''
            select
                v.schedule as schedule,
                v.employment as employment,
                count(*) as cnt
            from public.vacancies as v
            group by schedule, employment
            order by cnt desc
            '''

In [22]:
# результат запроса
schedule_employment = pd.read_sql_query(query_4_4, connection)
display(round(schedule_employment))

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

**Задание 4.5 - Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта.** 

In [23]:
# текст запроса
query_4_5 = f'''
            select
                v.experience as experience,
                count(*) as cnt
            from public.vacancies as v
            group by experience
            order by cnt
            '''

In [24]:
# результат запроса
experience_vacancies = pd.read_sql_query(query_4_5, connection)
display(experience_vacancies)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

### Дополнительный запрос 4.5.1 : 4.5.2 : 4.5.3  - анализ вакнсий с опытом работы более 6 лет.

In [34]:
# текст запроса - вывести наименование вакансий с опытом работы 6+ лет и их количество.
query_4_5_1 = f'''
            select
                v.name as name_vacancies,
                count(*) as cnt
            from public.vacancies as v
            where v.experience = 'Более 6 лет'
            group by v.name
            order by cnt desc
            limit 10
            '''

In [37]:
# результат запроса
experience_vacancies_6 = pd.read_sql_query(query_4_5_1, connection)
display(experience_vacancies_6)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name_vacancies  cnt
0                               Руководитель проекта   37
1                            Системный администратор   31
2                               Технический директор   28
3                                     Программист 1С   15
4                              Senior Data Scientist   14
5                                        Программист   11
6  Директор научно-исследовательского центра (химия)   10
7                              Senior Java Developer    9
8  Руководитель направления ТОиР Центра развития ...    8
9                        Senior System Administrator    8

***

In [42]:
# текст запроса - вывести распределение вакансий с опытом работы 6+ лет по городам и уровень СЗ.
query_4_5_2 = f'''
            select
                a.name as name_cities,
                count(*) as cnt,
                round(avg(salary_from)) as avg_salary_from,
                round(avg(salary_to)) as avg_salary_to
            from public.vacancies as v
                left join public.areas as a on a.id = v.area_id
            where v.experience = 'Более 6 лет'
            group by a.name
            order by cnt desc
            limit 10
            '''

In [43]:
# результат запроса
experience_vacancies_6_city = pd.read_sql_query(query_4_5_2, connection)
display(experience_vacancies_6_city)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name_cities  cnt  avg_salary_from  avg_salary_to
0           Москва  205         217389.0       273931.0
1           Алматы  101         117164.0       165683.0
2  Санкт-Петербург   74         174463.0       190000.0
3            Минск   66         118480.0       175265.0
4      Новосибирск   55         218100.0       292215.0
5       Нур-Султан   47         124462.0       114330.0
6  Нижний Новгород   42         193662.0       426808.0
7          Тбилиси   38         246068.0       317083.0
8          Ташкент   36         109015.0       145805.0
9       Красноярск   33         130088.0       115385.0

In [44]:
# текст запроса - вывести распределение вакансий с опытом работы 6+ лет по названию работодателей.
query_4_5_3 = f'''
            select
                e.name as name_employer,
                count(*) as cnt,
                round(avg(salary_from)) as avg_salary_from,
                round(avg(salary_to)) as avg_salary_to
            from public.vacancies as v
                left join public.employers as e on e.id = v.employer_id
            where v.experience = 'Более 6 лет'
            group by e.name
            order by cnt desc
            limit 10
            '''

In [45]:
# результат запроса
experience_vacancies_6_employer = pd.read_sql_query(query_4_5_3, connection)
display(experience_vacancies_6_employer)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name_employer  cnt  avg_salary_from  avg_salary_to
0             Bell Integrator   27              NaN            NaN
1                      Айсбит   17          80000.0            NaN
2  ЭФКО, Управляющая компания   15              NaN            NaN
3                  Volna.tech   13         116154.0       346154.0
4                     Playrix   12              NaN            NaN
5             Удоканская медь   11              NaN            NaN
6               Август, Фирма   10         500000.0            NaN
7               РОДЖИИ ЕВРОПА   10         278870.0       275345.0
8                 Восток.Софт    9         150000.0            NaN
9               Газпром нефть    8              NaN            NaN

### Выводы по детальному анализу вакансий

1. В базе представлена информация по 49197 вакансиям;
2. В 5 лидеров по количеству вакансий: Москва - 5333 (11%), Санкт-Петербург - 2851 (6%), Минск - 2112 (4%), Алматы - 1892 (4%). Из лидеров мы видим, что в базе представлены вакансии по РФ и странам СНГ;
3. По признаку - опыту работы, основной категорией являетс "От 1 года до 3 лет" - 26152 (53%) и "От 3 до 6 лет" - 14511 (29%). Очень низкое количество вакансий в категории опытных работников "Более 6 лет" - 1337 (3%);
4. Дополнительный исследования вакансий с опытом работы 6+ лет: 
    - к вакансиям с опытом работы "Более 6 лет" относятяся руководители проектов / направлений, директора (технические и прочие), лидеры / Senior и т.д. (запрос 4_5_1);
    - в разрезе городов - основная потребность представлена в городах миллиониках и столицах (запрос 4_5_2);
    - в разрезе работодателей информация представлена в запросе 4_5_3. Выводы сделать сложно, в основном не известные компании. ЗП в большей части не отображена (по договоренности).
5. Основные условия по рабочему графику и типу трудоуствройству: "Полный день" - "Полная занятость" - 35367 (72%);
6. Средний уровень ЗП по вакансиям в которых отражены предложения по ЗП ( 24073 / 49%) составляет от 71тыс.руб до 110тыс.руб. В разрезе признака "Опыт работы" средний уровень ЗП повышается - больше ЗП, выше ЗП (запрос 4_3_1).  

# Юнит 5. Анализ работодателей

**Задание 5.1 - Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.**

In [26]:
# текст запроса
query_5_1 = f'''
            select
                e.name as name_employers,
                count(*) as cnt
            from public.employers as e
            join public.vacancies as v on e.id = v.employer_id
            group by name_employers
            order by cnt desc
            limit 5
            '''

In [27]:
# результат запроса
top_5_employers = pd.read_sql_query(query_5_1, connection)
display(top_5_employers)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name_employers   cnt
0         Яндекс  1933
1     Ростелеком   491
2       Тинькофф   444
3           СБЕР   428
4  Газпром нефть   331

**Задание 5.2 - Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём. Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.**


In [45]:
# текст запроса
query_5_2 = f'''
            select
                a.name as name_area,
                count(distinct e.id) as cnt_employer,
                count(distinct v.id) as cnt_vacancies                
            from public.areas as a
                left join public.employers as e on a.id = e.area
                left join public.vacancies as v on v.area_id = a.id
            where v.id is null
            group by name_area
            order by cnt_employer desc
            limit 1
            '''

In [46]:
area_employer_vacancies = pd.read_sql_query(query_5_2, connection)
display(area_employer_vacancies)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name_area  cnt_employer  cnt_vacancies
0    Россия           410              0

**Задание 5.3 - Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.**


In [10]:
# текст запроса
query_5_3 = f'''
            select
                e.name as name_employer,
                count(distinct v.area_id) as cnt_area_public
            from public.employers as e
                left join public.vacancies as v on e.id = v.employer_id
            group by name_employer
            order by cnt_area_public desc
            limit 1
            '''

In [11]:
# результат запроса
public_area = pd.read_sql_query(query_5_3, connection)
display(public_area)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name_employer  cnt_area_public
0        Яндекс              181

**Задание 5.4 - Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности.** 

In [35]:
# текст запроса
query_5_4 = f'''
            select
                count(*) as cnt_employer_industries_null
            from public.employers as e
                left join public.employers_industries as ei on e.id = ei.employer_id
            where ei.industry_id is null
            
            '''

In [36]:
# результат запроса
employer_industries_null = pd.read_sql_query(query_5_4, connection)
display(employer_industries_null)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt_employer_industries_null
0                          8419

**Задание 5.5 - Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности.** 

In [44]:
# текст запроса
query_5_5 = f'''
            select
                e.name as name_employer,
                count(distinct ei.industry_id) as cnt_employer_industries
            from public.employers as e
                left join public.employers_industries as ei on e.id = ei.employer_id
            group by name_employer
            having count(distinct ei.industry_id) = 4
            offset 2 limit 1
            '''

In [45]:
# результат запроса
employer_industries_4 = pd.read_sql_query(query_5_5, connection)
display(employer_industries_4)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name_employer  cnt_employer_industries
0          2ГИС                        4

**Задание 5.6 - С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.**


In [46]:
# текст запроса
query_5_6 = f'''
            select
                --e.name
                count(e.name) as cnt_software_development
            from
                public.employers as e
                left join public.employers_industries as ei on e.id = ei.employer_id
                left join public.industries as i on ei.industry_id = i.id
            where i.name ilike '%Разработка программного обеспечения%'
            '''

In [47]:
# результат запроса
employers_software_development = pd.read_sql_query(query_5_6, connection)
display(employers_software_development)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt_software_development
0                      3553

**Задание 5.7 - Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.**

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [20]:
# код для получения списка городов-милионников
import urllib.parse as req
import pandas as pd

url_k = 'https://ru.wikipedia.org/wiki/Города-миллионеры_России#Список_городов-миллионеров'
path_common = req.urlparse(url_k)
url = path_common.scheme + '://' + req.quote(path_common.netloc) +  req.quote(path_common.path) + '#'+ req.quote(path_common.fragment)
table = pd.read_html(url)[1]

cities = tuple(table['Город'].unique())

In [25]:
# текст запроса
query_5_7 = f'''
            select
                a.name as name_city,
                count(distinct v.id) as cnt_vacancies
            from 
                public.areas as a
                left join public.vacancies as v on a.id = v.area_id
                left join public.employers as e on v.employer_id = e.id
            where e.name = 'Яндекс'
                and a.name in {cities}
            group by name_city
            
            union all
            
            select
                'Total'::text,
                count(distinct v.id) as cnt_vacancies
            from 
                public.areas as a
                left join public.vacancies as v on a.id = v.area_id
                left join public.employers as e on v.employer_id = e.id
            where e.name = 'Яндекс'
                and a.name in {cities}
                 
            order by cnt_vacancies
            '''

In [26]:
# результат запроса
area_vacancies = pd.read_sql_query(query_5_7, connection)
display(area_vacancies)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


name_city  cnt_vacancies
0              Омск             21
1         Челябинск             22
2        Красноярск             23
3         Волгоград             24
4             Пермь             25
5            Казань             25
6    Ростов-на-Дону             25
7               Уфа             26
8            Самара             26
9         Краснодар             30
10          Воронеж             32
11      Новосибирск             35
12  Нижний Новгород             36
13     Екатеринбург             39
14  Санкт-Петербург             42
15           Москва             54
16            Total            485

***

### Выводы по анализу работодателей

1. В базе представлена информация по 23501 работадателю;
2. Лидерами по количеству вакансий, представленных в базе, являются: Яндекс (информационные технологии) - 1933(4%),  (телекоммуникации) - 491(1%), Тинькофф (банковский сектор) - 444(0,9%), СБЕР (банковский сектор) - 428(0,9%), Газпромнефть (нефтяная компания) - 331 (0,7%);
3. В базе представлена информация по 294 наименованиям сфер деятельности работодателей. 8419(36%) компаний - работодателей не указали свою сферу деятельности. Максимальное количество указанных сфер деятельности составляет - 16 (2 компании: Модуль, Энергия). По 1 сфере указано у 6377 компаний(27%), 2 - 3602(15%), 3 - 2574(11%);
4. 3553(15%) компаний указали своей сферой деятельностью "Разработка программного обеспечения";
5. В базе указано 593 региона в которых 0 вакансий. В данную выборку входят страны (Россия, Казахстан, Беларусь, Сингапур и т.д), регионы / субьекты (Московская область, Ставропольский край и т.д). Как предложение - доработать таблицу и выделить отдельные столбцы: страна / область или субьект / город или населенный пункт.          

# Юнит 6. Предметный анализ

**Задание 6.1 - Сколько вакансий имеет отношение к данным?**

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [120]:
# текст запроса
query_6_1 = f'''
            select
                count(lower(v.name)) as cnt
            from
                public.vacancies as v
            where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
            '''

In [121]:
# результат запроса
vacancies_data = pd.read_sql_query(query_6_1, connection)
display(vacancies_data)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  1771

**Задание 6.2 - Сколько есть подходящих вакансий для начинающего дата-сайентиста?**
 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [309]:
# текст запроса
query_6_2 = f'''
            select
                count(v.name) as cnt
            from
                public.vacancies as v
            where (
                    lower(v.name) like '%data scientist%'
                    or lower(v.name) like '%data science%'
                    or lower(v.name) like '%исследователь данных%'
                    or (v.name ilike '%ML%' and not v.name ilike '%HTML%')
                    or lower(v.name) like '%machine learning%'
                    or lower(v.name) like '%машинн%обучен%'
                    )
                    and
                    (v.name ilike '%junior%' 
                    or v.experience ilike 'нет опыта' 
                    or v.employment ilike '%стажировка%'
                    )  
            '''


In [310]:
# результат запроса
vacancies_junior = pd.read_sql_query(query_6_2, connection)
display(vacancies_junior)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0   51

**Задание 6.3 - Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?**

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [97]:
query_6_3 = f'''
            select 
                count(*) as cnt
            from 
                public.vacancies as v  
            where
                (lower(v.name) like '%data scientist%'
                or lower(v.name) like '%data science%'            
                or v.name like '%ML%' and not v.name like '%HTML%'
                or lower(v.name) like '%исследователь данных%'
                or lower(v.name) like '%machine learning%'
                or lower(v.name) like '%машинн%обучен%')             
                and (lower(v.key_skills) like '%sql%' 
                or lower(v.key_skills) like '%postgres%'
                )     
            '''

In [98]:
vacancies_SQL_postgres = pd.read_sql_query(query_6_3, connection)
display(vacancies_SQL_postgres)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  201

In [96]:
# текст запроса - дополнительный материал (ответ не правильный)
query_6_3_1 = f'''
            select
                count(*) as cnt
            from
                public.vacancies as v
            where
                (v.name ~* 'data scien(ce|tist)|исследователь данных|ML|machine learning|машинн.обучен.' and v.name !~* 'HTML')
                and (v.key_skills ~* 'SQL' or v.key_skills ~* 'postgres') 
            '''

In [113]:
# результат запроса - дополнительный материал (ответ не правильный)
vacancies_SQL = pd.read_sql_query(query_6_3_1, connection)
display(vacancies_SQL)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  225

**Задание 6.4 - Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.**

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [110]:
query_6_4 = f'''
            select 
                count(*) as cnt
            from 
                public.vacancies as v  
            where
                (lower(v.name) like '%data scientist%'
                or lower(v.name) like '%data science%'            
                or v.name like '%ML%' and not v.name like '%HTML%'
                or lower(v.name) like '%исследователь данных%'
                or lower(v.name) like '%machine learning%'
                or lower(v.name) like '%машинн%обучен%')             
                and v.key_skills ilike '%Python%'     
            '''

In [108]:
# результат запроса
vacancies_python = pd.read_sql_query(query_6_4, connection)
display(vacancies_python)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  351

In [111]:
# текст запроса - дополнительный материал 
query_6_4_1 = f'''
            select
                count(*) as cnt
            from
                public.vacancies as v
            where(v.name ~* 'data scien(ce|tist)|исследователь данных|ML|machine learning|машинн.обучен.' and v.name !~* 'HTML')
            and (v.key_skills ~* 'Python')
            '''

In [112]:
# результат запроса - дополнительный материал
vacancies_python1 = pd.read_sql_query(query_6_4_1, connection)
display(vacancies_python1)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  351

**Задание 6.5 - Сколько ключевых навыков в среднем указывают в вакансиях для DS?**

Ответ округлите до двух знаков после точки-разделителя.

In [141]:
# текст запроса
query_6_5 = f'''
            select 
                replace(v.key_skills, CHR(9), ',') as key_skills_ds,
                --length(v.key_skills) as cnt1,
                --length(replace(v.key_skills, CHR(9), '')) as cnt2,
                (length(v.key_skills) - length(replace(v.key_skills, CHR(9), '')))+1 as cnt_key_skills_ds
            from 
                public.vacancies as v  
            where
                (lower(v.name) like '%data scientist%'
                or lower(v.name) like '%data science%'            
                or v.name like '%ML%' and not v.name like '%HTML%'
                or lower(v.name) like '%исследователь данных%'
                or lower(v.name) like '%machine learning%'
                or lower(v.name) like '%машинн%обучен%'     
                )  
            union all
            select 
                'Total (avg)'::text,
                round(avg((length(v.key_skills) - length(replace(v.key_skills, CHR(9), '')))+1),2) as cnt_key_skills_ds
            from 
                public.vacancies as v  
            where
                (lower(v.name) like '%data scientist%'
                or lower(v.name) like '%data science%'            
                or v.name like '%ML%' and not v.name like '%HTML%'
                or lower(v.name) like '%исследователь данных%'
                or lower(v.name) like '%machine learning%'
                or lower(v.name) like '%машинн%обучен%'     
                )
            '''

In [142]:
# результат запроса
key_skills_ds = pd.read_sql_query(query_6_5, connection)
display(key_skills_ds)

key_skills_ds  cnt_key_skills_ds
0                Python,SQL,Работа в команде,OpenCV,ML               5.00
1    Oracle Pl/SQL,Базы данных,Machine Learning,Маш...               9.00
2    Python,OpenCV,Tensorflow,Computer Vision,Нейро...               8.00
3                           Python,Flask,Linux,AWS,Git               5.00
4         Python,SQL,NLP,Scikit-learn,Machine Learning               5.00
..                                                 ...                ...
476               Python,Scikit-learn,Pandas,Git,Numpy               5.00
477  Qt,ООП,C++,MS Visual Studio,Разработка компьют...               6.00
478  SCALA,SQL,Математический анализ,Статистический...               7.00
479  Python,Linux,Kubernetes,AWS,Git,People Management               6.00
480                                        Total (avg)               6.41

[481 rows x 2 columns]

**Задание 6.6 - Напишите запрос, позволяющий вычислить, какую зарплату для DS в *среднем* указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*).**

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [224]:
# текст запроса
query_6_6 = f'''
            select
                v.experience as experience, 
                --v.salary_from as salary_from,
                --v.salary_to as salary_to,
                round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0) as avg_salary 
            from 
                public.vacancies as v
            where
                (lower(v.name) like '%data scientist%'
                or lower(v.name) like '%data science%'            
                or v.name like '%ML%' and not v.name like '%HTML%'
                or lower(v.name) like '%исследователь данных%'
                or lower(v.name) like '%machine learning%'
                or lower(v.name) like '%машинн%обучен%'     
                )
                and
                (v.salary_from is not null or v.salary_to is not null)  
            group by v.experience
            '''

In [225]:
# результат запроса
avg_salary_ds = pd.read_sql_query(query_6_6, connection)
display(avg_salary_ds)

experience  avg_salary
0           Нет опыта     74643.0
1  От 1 года до 3 лет    139675.0
2       От 3 до 6 лет    243115.0

***

### Выводы по предметному анализу

1. В приведенной базе отношение к "Data" имеет 1771(3.6%) вакансия, в том числе к "Data Science" - 536(1%) вакансий;
2. Для "Data Science" Junior подходит 51(9,5% от "Data") вакансия;
3. Среднее количество ключевых навыков в "Data Science" составляет 6,41, в том числе знание "Python" выделено у 351(65% от DS) вакансии, знание "SQL/Postgres" выделено у 201(37% от DS) вакансии;
4. Средний уровень ЗП в "Data Science" по признаку "Опыт работы" составляет: "нет опыта" - 75тыс.руб., " от 1 до 3 лет" - 140тыс.руб., "от 3 до 6 лет" - 243тыс.руб.  

# Общий вывод по проекту

1. База данных состоит из 5 таблиц в которых представлена информация о 49197 вакансиях в 23501 компаниях - работодателях, которые работают в 294 сферах деятельности и представлены в 1362 регионах (в том числе странах ближнего и дальнего зарубежья);

2. Наибольшее количество вакансий представлено в г.Москве, г.Санкт-Петербург, г.Минск и других столичных городах и городах с населением 1+ млн.чел. В базе присутсвуют регионы в которых вакансии не представлены - 0 и которые несут общие характеристики: страна, область / регион и т.д;

3. Крупнейшими работодателями (наибольшее количество вакансий) являются: Яндекс, Ростелеком, Тинькофф. Здесь надо проводить дополнительный анализ на предмет повторяющихся вакансий, с учетом того, что компании публикуют свои вакансии в нескольких регионах одновременно (Яндекс -181, Ростелеком - 152 и т.д.);

4. В основном требуется работники с опытом работы "от 1 до 3 лет" - 53% и "от 3 до 6 лет" - 29%. Основные предложения в части графика работы и трудоуствройства - "полный рабочий день - полная занятость" - 72%;

5. Ориентир по заработной плате указан у 49% вакансий, средний уровень ЗП составляет от 71 тыс.рублей до 110 тыс.рублей;

6. В части "Data Science" представлено около 1% вакансий. Ддополнительные исследования представлены в Юните - 7.

# Юнит 7. Дополнительный анализ в части "Data Science"

### Исследование 7.1 - Определим, сколько вакансий относится к "Data Science".

In [161]:
# текст запроса
query_7_1 = f'''
            select
                count(v.name) as cnt
            from
                public.vacancies as v
            where (
                    lower(v.name) like '%data scientist%'
                    or lower(v.name) like '%data science%'
                    or lower(v.name) like '%исследователь данных%'
                    or (v.name ilike '%ML%' and not v.name ilike '%HTML%')
                    or lower(v.name) like '%machine learning%'
                    or lower(v.name) like '%машинн%обучен%'
                    )
            '''


In [162]:
# результат запроса
vacancies_data_science = pd.read_sql_query(query_7_1, connection)
display(vacancies_data_science)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0  536

**Комментарии:**

1. К DS относится 536 вакнсий, что составляет 1% от базы. 

### Исследование 7.2 - Определим, по скольким вакансиям в "Data Science" есть информация по уровню ЗП ("от" или "до").

In [163]:
# текст запроса
query_7_2 = f'''
            select
                count(v.name) as cnt
            from
                public.vacancies as v
            where (
                    lower(v.name) like '%data scientist%'
                    or lower(v.name) like '%data science%'
                    or lower(v.name) like '%исследователь данных%'
                    or (v.name ilike '%ML%' and not v.name ilike '%HTML%')
                    or lower(v.name) like '%machine learning%'
                    or lower(v.name) like '%машинн%обучен%'
                    )
                    and 
                    (v.salary_from is not null and v.salary_to is not null)
            '''

In [164]:
# результат запроса
vacancies_DS_with_salary = pd.read_sql_query(query_7_2, connection)
display(vacancies_DS_with_salary)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


cnt
0   39

**Комментарии:**

1. Из выборки 536 вакнсий по 39 позициям указан ориентировочный уровень ЗП что составляет, что составляет 0,7% от выборки и не может служить (серьезным) ориентиром для анализа. 

### Исследование 7.3 - Определим в каких регионах представлены ваканси "Data Science" и средний уровень ЗП.

In [16]:
# текст запроса
query_7_3 = f'''
            select
                a.name as city_name,
                count(a.name) as cnt,
                round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0) as avg_salary
            from
                public.vacancies as v
                left join public.areas as a on v.area_id = a.id
            where (
                    lower(v.name) like '%data scientist%'
                    or lower(v.name) like '%data science%'
                    or lower(v.name) like '%исследователь данных%'
                    or (v.name ilike '%ML%' and not v.name ilike '%HTML%')
                    or lower(v.name) like '%machine learning%'
                    or lower(v.name) like '%машинн%обучен%'
                    )
            group by a.name
            order by cnt desc
            limit 10
            '''

In [17]:
# результат запроса
vacancies_DS_city = pd.read_sql_query(query_7_3, connection)
display(vacancies_DS_city)

city_name  cnt  avg_salary
0           Москва  225    214000.0
1  Санкт-Петербург   69    183150.0
2      Новосибирск   23    166333.0
3           Алматы   22    120013.0
4  Нижний Новгород   20         NaN
5           Казань   19    161833.0
6   Ростов-на-Дону   11    241667.0
7            Томск   11         NaN
8            Минск   11    314785.0
9            Пермь    8    116667.0

**Комментарии:**

1. Лидером по количеству вакансий в DS является Москва. Уровень ЗП в зависимости от города / региона меняется, но с учетом малого количества заполненных данных по ЗП нужно относится к нему с осторожностью. Так же необходимо проводить дополнительный анализ по уровню вакансии (Junior, Middle, Senior / Lead) 

### Исследование 7.4 - Определим в каких компаниях представлены ваканси "Data Science".

In [20]:
# текст запроса
query_7_4 = f'''
            select
                e.name as employers_name,
                count(a.name) as cnt,
                round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to)), 0) as avg_salary
            from
                public.vacancies as v
                left join public.areas as a on v.area_id = a.id
                left join public.employers as e on v.employer_id = e.id 
                
            where (
                    lower(v.name) like '%data scientist%'
                    or lower(v.name) like '%data science%'
                    or lower(v.name) like '%исследователь данных%'
                    or (v.name ilike '%ML%' and not v.name ilike '%HTML%')
                    or lower(v.name) like '%machine learning%'
                    or lower(v.name) like '%машинн%обучен%'
                    )
            group by e.name
            order by cnt desc
            limit 10
            '''

In [21]:
# результат запроса
vacancies_DS_employers = pd.read_sql_query(query_7_4, connection)
display(vacancies_DS_employers)

employers_name  cnt  avg_salary
0                   СБЕР   39         NaN
1        Bell Integrator   25         NaN
2         Банк ВТБ (ПАО)   18         NaN
3                     VK   15         NaN
4  Positive Technologies   11         NaN
5                 Яндекс    9         NaN
6         EvenBet Gaming    9         NaN
7                МегаФон    8         NaN
8               Andersen    7    268863.0
9     Бэнкс Софт Системс    6         NaN

**Комментарии:**

1. Лидером по количеству вакансий в DS является СБЕР. Уровень ЗП у 9 из 10 выведенных компаний не указан. Малое количество вакансий у Яндекса, при том, что самое большое количество вакансий - 1933 (есть предположение, что вакансии повторяющиеся в разных регионах).

### Исследование 7.5 - Определим вакансии уровня Middle "Data Science", выделим ключевые навыки и требования по опыту работы.

In [22]:
# текст запроса
query_7_5 = f'''
            select
                v.name as vacancies_name,
                v.key_skills as key_skills_middle,
                v.experience as experience_middle
            from
                public.vacancies as v
                
            where (
                    lower(v.name) like '%data scientist%'
                    or lower(v.name) like '%data science%'
                    or lower(v.name) like '%исследователь данных%'
                    or (v.name ilike '%ML%' and not v.name ilike '%HTML%')
                    or lower(v.name) like '%machine learning%'
                    or lower(v.name) like '%машинн%обучен%'
                    )
                    and
                    (v.name not ilike '%junior%' 
                    or v.experience not ilike '%нет опыта%' 
                    or v.employment not ilike '%стажировка%'
                    )
                    and v.name ilike '%middle%' 
                    and v.name not ilike '%senior%'
            '''

In [23]:
# результат запроса
vacancies_DSM = pd.read_sql_query(query_7_5, connection)
display(vacancies_DSM)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


vacancies_name  \
0                               Middle Data Scientist   
1                     Data Scientist (junior+/middle)   
2                        Junior/Middle Data Scientist   
3                    Middle Data Scientist (Big Data)   
4                             Data Scientist (middle)   
5                               Middle Data Scientist   
6                               Middle Data Scientist   
7                               Middle Data Scientist   
8                               Middle Data Scientist   
9             Middle Data Scientist (Computer Vision)   
10   Middle data scientist (Рекомендательные системы)   
11               Data Scientist/ML Engineer (Middle+)   
12                              Middle Data Scientist   
13        Data scientist (Junior/Middle) в консалтинг   
14                            Data Scientist (Middle)   
15               Middle ML / Data Science разработчик   
16  Программист Python/ Data Scientist (junior / m...   
17                   Middle MLE (Big Data Department)   
18                   Middle MLE (Big Data Department)   
19                   Middle Data Scientist (Big Data)   
20                            Data Scientist (Middle)   
21                            Data scientist (middle)   
22                              Middle Data Scientist   
23                    Data scientist (Juniour-Middle)   
24                            Data scientist (Middle)   
25                              Middle Data Scientist   
26       Middle Data Scientist (клиентская аналитика)   
27       Middle Data Scientist (клиентская аналитика)   
28          Data Scientist (Computer Vision) / Middle   
29                                 Middle ML Engineer   
30                                 Middle ML Engineer   

                                    key_skills_middle   experience_middle  
0         Python\tLearning to Rank\tData Science\tNLP       От 3 до 6 лет  
1   Python\tМатематическая статистика\tOracle Pl/S...  От 1 года до 3 лет  
2                                       Linux\tPython  От 1 года до 3 лет  
3                     Python\tSQL\tGit\tLinux\tHadoop  От 1 года до 3 лет  
4   Python\tSQL\tData Mining\tC++\tМатематический ...  От 1 года до 3 лет  
5         Python\tLearning to Rank\tData Science\tNLP       От 3 до 6 лет  
6         Python\tLearning to Rank\tData Science\tNLP       От 3 до 6 лет  
7         Python\tLearning to Rank\tData Science\tNLP       От 3 до 6 лет  
8         Python\tLearning to Rank\tData Science\tNLP       От 3 до 6 лет  
9   OpenCV\tPyTorch\tPython\tАнглийский язык\tРабо...  От 1 года до 3 лет  
10                                               None  От 1 года до 3 лет  
11  SCALA\tJava\tSQL\tАнализ данных\tPython\tElast...  От 1 года до 3 лет  
12  Python\tSQL\tHadoop\tMachine Learning\tАнглийс...       От 3 до 6 лет  
13                                         Python\tML           Нет опыта  
14  Python\tSQL\tSpark\tУправление проектами\tScru...       От 3 до 6 лет  
15  NLP\tComputer Vision\tData Analysis\tData Scie...  От 1 года до 3 лет  
16         Python\tDjango Framework\tSQL\tXML\tMS SQL           Нет опыта  
17                 Python\tGit\tSQL\tScrum\tML\tSpark  От 1 года до 3 лет  
18                 Python\tGit\tSQL\tScrum\tML\tSpark  От 1 года до 3 лет  
19                                   Python\tSQL\tGit  От 1 года до 3 лет  
20  Python\tMachine Learning\tSpark\tDocker\tMS Po...       От 3 до 6 лет  
21  Математическое моделирование\tMATLAB\tМатемати...  От 1 года до 3 лет  
22        Python\tLearning to Rank\tData Science\tNLP       От 3 до 6 лет  
23  Python\tSQL\tMS PowerPoint\tАналитическое мышл...  От 1 года до 3 лет  
24  Анализ данных\tMS PowerPoint\tPandas\tPython\t...       От 3 до 6 лет  
25  Python\tSQL\tData Mining\tMachine Learning\tАн...  От 1 года до 3 лет  
26  Python\tSQL\tАнглийский язык\tМатематическое м...  От 1 года до 3 лет  
27  Python\tSQL\tАнглийский язык\tМатематическое м...  От 1 года до 3 лет  
28  Python

**Комментарии:**

1. В задание 6.2 было определено, что к Junior можно отнести 51 вакансию. Согласно запроса выведено 27 позиций по которым есть прямая ссылка на уровень Middle;
2. В требования по опыту работы разделились примерно по полам "от 1 до 3 лет" и "от 3 до 6 лет".

### Исследование 7.6 - Определим вакансии уровня Senior "Data Science", выделим ключевые навыки и требования по опыту работы.

In [24]:
# текст запроса
query_7_6 = f'''
            select
                v.name as vacancies_name,
                v.key_skills as key_skills_middle,
                v.experience as experience_middle
            from
                public.vacancies as v
                
            where (
                    lower(v.name) like '%data scientist%'
                    or lower(v.name) like '%data science%'
                    or lower(v.name) like '%исследователь данных%'
                    or (v.name ilike '%ML%' and not v.name ilike '%HTML%')
                    or lower(v.name) like '%machine learning%'
                    or lower(v.name) like '%машинн%обучен%'
                    )
                    and
                    (v.name not ilike '%junior%' 
                    or v.experience not ilike '%нет опыта%' 
                    or v.employment not ilike '%стажировка%'
                    )
                    --and v.name not ilike '%middle%' 
                    and v.name ilike '%senior%'
            '''

In [25]:
# результат запроса
vacancies_DSS = pd.read_sql_query(query_7_6, connection)
display(vacancies_DSS)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


vacancies_name  \
0                            AI/ML Senior Engineer   
1                              PHP Senior/TeamLead   
2                            AI/ML Senior Engineer   
3   Математик, Junior/Middle/Senior Data Scientist   
4                      Senior Developer (teamlead)   
..                                             ...   
66                                 TeamLead/Senior   
67                     Senior Developer (teamlead)   
68                   Middle+/Senior Data Scientist   
69     Системный аналитик /Middle+/Senior/TeamLead   
70  Frontend разработчик, VueJS (Senior), TeamLead   

                                    key_skills_middle   experience_middle  
0                      Python\tFlask\tLinux\tAWS\tGit       От 3 до 6 лет  
1   Разработка ПО\tJavaScript\tSQL\tMySQL\tВеб-про...  От 1 года до 3 лет  
2                      Python\tFlask\tLinux\tAWS\tGit       От 3 до 6 лет  
3   Python\tMATLAB\tМатематическое моделирование\t...  От 1 года до 3 лет  
4   Git\tООП\tPostgreSQL\tBPMN\tBackend\tРабота с ...         Более 6 лет  
..                                                ...                 ...  
66                  Java\tGithub\tOpenDNS\tPostgreSQL       От 3 до 6 лет  
67  Git\tООП\tPostgreSQL\tBPMN\tBackend\tРабота с ...         Более 6 лет  
68  SCALA\tSQL\tМатематический анализ\tСтатистичес...       От 3 до 6 лет  
69  UML\tSQL\tBPMN\tSOAP\tREST\tПостановка задач р...       От 3 до 6 лет  
70  Git\tJavaScript\tHTML\tCSS\tVue.js\tVue\tREST ...  От 1 года до 3 лет  

[71 rows x 3 columns]

**Комментарии:**

1. Согласно запроса выведено 71 позиция по которым есть прямая ссылка на уровень Senior;
2. В требования по опыту работы разделились примерно по полам "от 3 до 6 лет" и "более 6 лет";
3. С учетом ранее определенного количества Junior и Middle остается большое количество вакансий в которых четко не определен уровень (скорей всего между Middle и Senior).

### Исследование 7.7 - Определим прочие вакансии "Data Science", выделим ключевые навыки и требования по опыту работы.

In [26]:
# текст запроса
query_7_7 = f'''
            select
                v.name as vacancies_name,
                v.key_skills as key_skills_middle,
                v.experience as experience_middle
            from
                public.vacancies as v
                
            where (
                    lower(v.name) like '%data scientist%'
                    or lower(v.name) like '%data science%'
                    or lower(v.name) like '%исследователь данных%'
                    or (v.name ilike '%ML%' and not v.name ilike '%HTML%')
                    or lower(v.name) like '%machine learning%'
                    or lower(v.name) like '%машинн%обучен%'
                    )
                    and v.name not ilike '%junior%' 
                    and v.experience not ilike '%нет опыта%' 
                    and v.employment not ilike '%стажировка%'
                    and v.name not ilike '%middle%' 
                    and v.name not ilike '%senior%'
            '''

In [27]:
# результат запроса
vacancies_DS_other = pd.read_sql_query(query_7_7, connection)
display(vacancies_DS_other)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


vacancies_name  \
0                                       Data Scientist   
1                               Разработчик C++/Qt/QML   
2                               QA Automation TeamLead   
3                                       Data Scientist   
4                                       Data Scientist   
..                                                 ...   
385  Data scientist / Исследователь-программист данных   
386                              C++ разработчик (QML)   
387                                   Специалист ML/DS   
388   Data Scientist в области рекомендательных систем   
389                  ML Engineer Team Lead (to Cyprus)   

                                     key_skills_middle   experience_middle  
0    Oracle Pl/SQL\tБазы данных\tMachine Learning\t...  От 1 года до 3 лет  
1                           C++\tQt\tVoIP\tTCP/IP\tQML  От 1 года до 3 лет  
2    Atlassian Jira\tФункциональное тестирование\tS...       От 3 до 6 лет  
3     Python\tSQL\tNLP\tScikit-learn\tMachine Learning  От 1 года до 3 лет  
4     Python\tSQL\tNLP\tScikit-learn\tMachine Learning  От 1 года до 3 лет  
..                                                 ...                 ...  
385  Python\tАналитическое мышление\tАнализ данных\...  От 1 года до 3 лет  
386  Qt\tООП\tC++\tMS Visual Studio\tРазработка ком...       От 3 до 6 лет  
387  Python\tNumpy\tPyTorch\tPandas\tМатематический...       От 3 до 6 лет  
388           Python\tScikit-learn\tPandas\tGit\tNumpy       От 3 до 6 лет  
389  Python\tLinux\tKubernetes\tAWS\tGit\tPeople Ma...       От 3 до 6 лет  

[390 rows x 3 columns]

**Комментарии:**

1. Согласно запроса выведено 390(73%) позиций по которым не определен уровень вакансии;
2. Можно предположить, что в компаниях нет четкого понимания / определения границ между Middle и Senior или она размыта;
3. 

### Исследование 7.8 Распределение прочих вакансий по опыту работы в "Data Science".

In [12]:
# текст запроса
query_7_9 = f'''
            select
                v.experience as experience,
                count(*) as cnt
            from
                public.vacancies as v
                
            where (
                    lower(v.name) like '%data scientist%'
                    or lower(v.name) like '%data science%'
                    or lower(v.name) like '%исследователь данных%'
                    or (v.name ilike '%ML%' and not v.name ilike '%HTML%')
                    or lower(v.name) like '%machine learning%'
                    or lower(v.name) like '%машинн%обучен%'
                    )
                    and v.name not ilike '%junior%' 
                    and v.experience not ilike '%нет опыта%' 
                    and v.employment not ilike '%стажировка%'
                    and v.name not ilike '%middle%' 
                    and v.name not ilike '%senior%'
            group by v.experience
            order by cnt desc
            --limit 5
            '''

In [13]:
# результат запроса
vacancies_DS_exp = pd.read_sql_query(query_7_9, connection)
display(vacancies_DS_exp)

/Users/alexander.v.varlamov/opt/anaconda3/lib/python3.9/site-packages/pandas/io/sql.py:762: UserWarning: pandas only support SQLAlchemy connectable(engine/connection) ordatabase string URI or sqlite3 DBAPI2 connectionother DBAPI2 objects are not tested, please consider using SQLAlchemy
  warnings.warn(


experience  cnt
0       От 3 до 6 лет  183
1  От 1 года до 3 лет  182
2         Более 6 лет   25

**Комментарии:**

1. Согласно запроса выведено 390(73%) позиций по которым не определен уровень вакансии;
2. Можно предположить, что в компаниях нет четкого понимания / определения границ между Middle и Senior;
3. Остается 12 вакансий не учтенных в выборке.

In [28]:
connection.close()